packages/dcos-integration-test/extra/test_metrics.py

```diff
@@ -75,9 +75,19 @@ def wait_for_package_teardown():
             assert state_response.status_code == 200
             state = state_response.json()
 
+            # Rarely, the framework will continue to show up in 'frameworks' instead of
+            # 'completed_frameworks', even after teardown. To avoid this causing a test
+            # failure, if the framework continues to show up in 'frameworks', we instead
+            # check if there are any running tasks.
+            completed = True
             for framework in state['frameworks']:
                 if framework['name'] == framework_name:
-                    raise Exception('Framework {} still running'.format(framework_name))
+                    completed = False
+                    if len(framework['tasks']) == 0:
+                        return
+            if completed:
+                return
+            raise Exception('Framework {} still running'.format(framework_name))
```

Comment:
Adding the `completed` variable complicates this `for` loop to the point where I had a hard time following it. Let's clarify this by building a `dict` of frameworks from the response, and then making assertions on that data structure:

```python
frameworks = {f['name']: f for f in state['frameworks']}
assert framework_name not in frameworks or len(frameworks[framework_name]['tasks']) == 0
```

Reply:
whoaaa dict comprehensions 😱